From 521508975edc13442fef4db6611b142d8ca15127 Mon Sep 17 00:00:00 2001
From: Baji Patthan <bpatth@codeaurora.org>
Date: Tue, 5 May 2015 19:30:18 -0700
Subject: [PATCH 05/22] ARM: dts: msm: Add initial device tree for SBC APQ8016

Add the device tree files necessary to support the SBC8016 platform.

Change-Id: I3ca144ecc40afb79c569daed21c309bbc7c870ef
Signed-off-by: Baji Patthan <bpatth@codeaurora.org>
---
 Documentation/devicetree/bindings/arm/msm/msm.txt  | 218 +++++++
 arch/arm/boot/dts/qcom/Makefile                    |   1 +
 .../boot/dts/qcom/apq8016-camera-sensor-sbc.dtsi   | 271 +++++++++
 arch/arm/boot/dts/qcom/apq8016-sbc.dts             |  25 +
 arch/arm/boot/dts/qcom/apq8016-sbc.dtsi            | 663 +++++++++++++++++++++
 arch/arm/boot/dts/qcom/apq8016.dtsi                |  33 +
 arch/arm/boot/dts/qcom/msm8916-pinctrl.dtsi        |  54 +-
 7 files changed, 1264 insertions(+), 1 deletion(-)
 create mode 100644 Documentation/devicetree/bindings/arm/msm/msm.txt
 create mode 100644 arch/arm/boot/dts/qcom/apq8016-camera-sensor-sbc.dtsi
 create mode 100644 arch/arm/boot/dts/qcom/apq8016-sbc.dts
 create mode 100644 arch/arm/boot/dts/qcom/apq8016-sbc.dtsi
 create mode 100644 arch/arm/boot/dts/qcom/apq8016.dtsi

diff --git a/Documentation/devicetree/bindings/arm/msm/msm.txt b/Documentation/devicetree/bindings/arm/msm/msm.txt
new file mode 100644
index 0000000..3440e64
--- /dev/null
+++ b/Documentation/devicetree/bindings/arm/msm/msm.txt
@@ -0,0 +1,218 @@
+* Qualcomm Technologies, Inc. MSM
+
+MSM uses a combination of DTS and DTSI files to describe the hardware on various
+SoCs and boards. Typically, a SoC-specific DTSI file describes the devices
+present on a given SoC, and a board-specific DTSI file describes the devices
+external to the SoC, although some targets may follow a more simplified
+approach. Additionally, the SoC-specific DTSI files may further consist of a
+base chip-specific file and a version-specific DTSI file, to facilitate reuse
+of device definitions among multiple revisions of the same SoC.
+
+Required properties:
+- compatible: Every device present on the MSM SoC shall have a 'qcom,' prefix
+  in its compatible string
+
+Example:
+restart@fc4ab000 {
+	compatible = "qcom,pshold";
+	reg = <0xfc4ab000 0x4>;
+};
+
+
+* Compatible strings:
+
+SoCs:
+
+- APQ8016
+  compatible = "qcom,apq8016"
+
+- APQ8026
+  compatible = "qcom,apq8026"
+
+- APQ8074
+  compatible = "qcom,apq8074"
+
+- APQ8084
+  compatible = "qcom,apq8084"
+
+- APQ8094
+  compatible = "qcom,apq8094"
+
+- MDM9630
+  compatible = "qcom,mdm9630"
+
+- MSM8226
+  compatible = "qcom,msm8226"
+
+- MSM8610
+  compatible = "qcom,msm8610"
+
+- MSM8909
+  compatible = "qcom,msm8909"
+
+- MSM8916
+  compatible = "qcom,msm8916"
+
+- APQ8016
+  compatible = "qcom,apq8016"
+
+- MSM8936
+  compatible = "qcom,msm8936"
+
+- MSM8960
+  compatible = "qcom,msm8960"
+
+- MSM8992
+  compatible = "qcom,msm8992"
+
+- MSM8994
+  compatible = "qcom,msm8994"
+
+- MSM8952
+  compatible = "qcom,msm8952"
+
+- MDM9640
+  compatible = "qcom,mdm9640"
+
+- VPIPA
+  compatible = "qcom,msmvpipa"
+
+- FSM9010
+  compatible = "qcom,fsm9010"
+
+- FSM9910
+  compatible = "qcom,fsm9900"
+
+- MSM8929
+  compatible = "qcom,msm8929"
+
+Generic board variants:
+
+- CDP device:
+  compatible = "qcom,cdp"
+
+- MTP device:
+  compatible = "qcom,mtp"
+
+- FLUID device:
+  compatible = "qcom,fluid"
+
+- LIQUID device:
+  compatible = "qcom,liquid"
+
+- Dragonboard device:
+  compatible = "qcom,dragonboard"
+
+- SURF device:
+  compatible = "qcom,surf"
+
+- QRD device:
+  compatible = "qcom,qrd"
+
+- SBC device:
+  compatible = "qcom,sbc"
+
+- Simulator device:
+  compatible = "qcom,sim"
+
+- RUMI device:
+  compatible = "qcom,rumi"
+
+
+
+Boards (SoC type + board variant):
+
+compatible = "qcom,apq8016"
+compatible = "qcom,apq8026-cdp"
+compatible = "qcom,apq8026-mtp"
+compatible = "qcom,apq8026-xpm"
+compatible = "qcom,apq8074-cdp"
+compatible = "qcom,apq8074-dragonboard"
+compatible = "qcom,apq8074-liquid"
+compatible = "qcom,apq8084-cdp"
+compatible = "qcom,apq8084-liquid"
+compatible = "qcom,apq8084-mtp"
+compatible = "qcom,apq8084-sbc"
+compatible = "qcom,apq8094-cdp"
+compatible = "qcom,apq8094-fluid"
+compatible = "qcom,apq8094-liquid"
+compatible = "qcom,apq8094-mtp"
+compatible = "qcom,apq8094-dragonboard"
+compatible = "qcom,mdm9630-cdp"
+compatible = "qcom,mdm9630-mtp"
+compatible = "qcom,mdm9630-sim"
+compatible = "qcom,msm8226-cdp"
+compatible = "qcom,msm8226-fluid"
+compatible = "qcom,msm8226-mtp"
+compatible = "qcom,msm8226-qrd"
+compatible = "qcom,msm8226-sim"
+compatible = "qcom,msm8610-cdp"
+compatible = "qcom,msm8610-mtp"
+compatible = "qcom,msm8610-qrd"
+compatible = "qcom,msm8610-rumi"
+compatible = "qcom,msm8610-sim"
+compatible = "qcom,msm8660-surf"
+compatible = "qcom,msm8909-cdp"
+compatible = "qcom,msm8909-mtp"
+compatible = "qcom,msm8909-qrd"
+compatible = "qcom,msm8909-rumi"
+compatible = "qcom,msm8909-sim"
+compatible = "qcom,msm8916-cdp"
+compatible = "qcom,msm8916-mtp"
+compatible = "qcom,msm8916-qrd-skuh"
+compatible = "qcom,msm8916-qrd-skuhf"
+compatible = "qcom,msm8916-qrd-skui"
+compatible = "qcom,msm8916-qrd-skuic"
+compatible = "qcom,msm8916-qrd-skuid"
+compatible = "qcom,msm8916-qrd-skut1"
+compatible = "qcom,msm8916-rumi"
+compatible = "qcom,msm8916-sim"
+compatible = "qcom,apq8016-sbc"
+compatible = "qcom,msm8926-cdp"
+compatible = "qcom,msm8926-mtp"
+compatible = "qcom,msm8926-qrd"
+compatible = "qcom,msm8936-cdp"
+compatible = "qcom,msm8936-mtp"
+compatible = "qcom,msm8939-cdp"
+compatible = "qcom,msm8939-mtp"
+compatible = "qcom,msm8939-qrd-skuk"
+compatible = "qcom,msm8939-qrd-skul"
+compatible = "qcom,msm8939-rumi"
+compatible = "qcom,msm8939-sim"
+compatible = "qcom,msm8960-cdp"
+compatible = "qcom,msm8974-cdp"
+compatible = "qcom,msm8974-fluid"
+compatible = "qcom,msm8974-liquid"
+compatible = "qcom,msm8974-mtp"
+compatible = "qcom,msm8974-rumi"
+compatible = "qcom,msm8974-sim"
+compatible = "qcom,msm8992-cdp"
+compatible = "qcom,msm8992-mtp"
+compatible = "qcom,msm8992-rumi"
+compatible = "qcom,msm8992-sim"
+compatible = "qcom,msm8994-cdp"
+compatible = "qcom,msm8994-fluid"
+compatible = "qcom,msm8994-liquid"
+compatible = "qcom,msm8994-mtp"
+compatible = "qcom,msm8994-rumi"
+compatible = "qcom,msm8994-sim"
+compatible = "qcom,msm8952-rumi"
+compatible = "qcom,msm8952-sim"
+compatible = "qcom,msmterbium-rumi"
+compatible = "qcom,msmterbium-sim"
+compatible = "qcom,msmterbium-cdp"
+compatible = "qcom,msmterbium-mtp"
+compatible = "qcom,mdm9640-cdp"
+compatible = "qcom,mdm9640-mtp"
+compatible = "qcom,mdm9640-rumi"
+compatible = "qcom,mdm9640-sim"
+compatible = "qcom,fsm9010-cdp"
+compatible = "qcom,fsm9010-mtp"
+compatible = "qcom,fsm9900-cdp"
+compatible = "qcom,fsm9900-mtp"
+compatible = "qcom,msmvpipa-sim"
+compatible = "qcom,msm8929-cdp"
+compatible = "qcom,msm8929-mtp"
+compatible = "qcom,msm8929-qrd"
+compatible = "qcom,msm8929-qrd-skul"
+compatible = "qcom,msm8929-qrd-skulb"
diff --git a/arch/arm/boot/dts/qcom/Makefile b/arch/arm/boot/dts/qcom/Makefile
index 277e196..3face4f 100644
--- a/arch/arm/boot/dts/qcom/Makefile
+++ b/arch/arm/boot/dts/qcom/Makefile
@@ -58,6 +58,7 @@ dtb-$(CONFIG_ARCH_MSM8916) += msm8916-sim.dtb \
 	msm8916-qrd-skut1.dtb \
 	apq8016-mtp.dtb \
 	apq8016-mtp-smb1360.dtb \
+	apq8016-sbc.dtb \
 	msm8939-sim.dtb \
 	msm8939-rumi.dtb \
 	msm8939-qrd-skuk.dtb \
diff --git a/arch/arm/boot/dts/qcom/apq8016-camera-sensor-sbc.dtsi b/arch/arm/boot/dts/qcom/apq8016-camera-sensor-sbc.dtsi
new file mode 100644
index 0000000..1bd1862
--- /dev/null
+++ b/arch/arm/boot/dts/qcom/apq8016-camera-sensor-sbc.dtsi
@@ -0,0 +1,271 @@
+/*
+ * Copyright (c) 2015, The Linux Foundation. All rights reserved.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 and
+ * only version 2 as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ */
+
+&soc {
+};
+
+&cci {
+
+	actuator0: qcom,actuator@0 {
+		cell-index = <0>;
+		reg = <0x0>;
+		compatible = "qcom,actuator";
+		qcom,cci-master = <0>;
+		cam_vaf-supply = <&vph_pwr_vreg>;
+		qcom,cam-vreg-name = "cam_vaf";
+		qcom,cam-vreg-min-voltage = <2800000>;
+		qcom,cam-vreg-max-voltage = <2800000>;
+		qcom,cam-vreg-op-mode = <80000>;
+	};
+
+	eeprom0: qcom,eeprom@20 {
+		cell-index = <0>;
+		reg = <0x20>;
+		qcom,eeprom-name = "sonyimx135";
+		compatible = "qcom,eeprom";
+		qcom,slave-addr = <0x20>;
+		qcom,cci-master = <0>;
+		qcom,num-blocks = <16>;
+
+		qcom,page0 = <1 0x3b02 2 0x10 1 5>;
+		qcom,pageen0 = <1 0x3b00 2 0x01 1 5>;
+		qcom,poll0 = <1 0x3b01 2 0x01 1 5>;
+		qcom,mem0 = <8 0x3b04 2 0 1 0>;
+
+		qcom,page1 = <1 0x3b02 2 0x0F 1 5>;
+		qcom,pageen1 = <1 0x3b00 2 0x01 1 5>;
+		qcom,poll1 = <1 0x3b01 2 0x01 1 5>;
+		qcom,mem1 = <8 0x3b04 2 0 1 0>;
+
+		qcom,page2 = <1 0x3b02 2 0x0E 1 5>;
+		qcom,pageen2 = <1 0x3b00 2 0x01 1 5>;
+		qcom,poll2 = <1 0x3b01 2 0x01 1 5>;
+		qcom,mem2 = <8 0x3b04 2 0 1 0>;
+
+		qcom,page3 = <1 0x3b02 2 0x0D 1 5>;
+		qcom,pageen3 = <1 0x3b00 2 0x01 1 5>;
+		qcom,poll3 = <1 0x3b01 2 0x01 1 5>;
+		qcom,mem3 = <8 0x3b04 2 0 1 0>;
+
+		qcom,page4 = <1 0x3b02 2 0x12 1 5>;
+		qcom,pageen4 = <1 0x3b00 2 0x01 1 5>;
+		qcom,poll4 = <1 0x3b01 2 0x01 1 5>;
+		qcom,mem4 = <7 0x3b09 2 0 1 0>;
+
+		qcom,page5 = <1 0x3b02 2 0x11 1 5>;
+		qcom,pageen5 = <1 0x3b00 2 0x01 1 5>;
+		qcom,poll5 = <1 0x3b01 2 0x01 1 5>;
+		qcom,mem5 = <7 0x3b09 2 0 1 0>;
+
+		qcom,page6 = <1 0x3b02 2 0x10 1 5>;
+		qcom,pageen6 = <1 0x3b00 2 0x01 1 5>;
+		qcom,poll6 = <1 0x3b01 2 0x01 1 5>;
+		qcom,mem6 = <7 0x3b09 2 0 1 0>;
+
+		qcom,page7 = <1 0x3b02 2 0x09 1 5>;
+		qcom,pageen7 = <1 0x3b00 2 0x01 1 5>;
+		qcom,poll7 = <1 0x3b01 2 0x01 1 5>;
+		qcom,mem7 = <7 0x3b09 2 0 1 0>;
+
+		qcom,page8 = <1 0x3b02 2 0x01 1 5>;
+		qcom,pageen8 = <1 0x3b00 2 0x01 1 5>;
+		qcom,poll8 = <1 0x3b01 2 0x01 1 5>;
+		qcom,mem8 = <64 0x3b04 2 0 1 0>;
+
+		qcom,page9 = <1 0x3b02 2 0x02 1 5>;
+		qcom,pageen9 = <1 0x3b00 2 0x01 1 5>;
+		qcom,poll9 = <1 0x3b01 2 0x01 1 5>;
+		qcom,mem9 = <64 0x3b04 2 0 1 0>;
+
+		qcom,page10 = <1 0x3b02 2 0x03 1 5>;
+		qcom,pageen10 = <1 0x3b00 2 0x01 1 5>;
+		qcom,poll10 = <1 0x3b01 2 0x01 1 5>;
+		qcom,mem10 = <64 0x3b04 2 0 1 0>;
+
+		qcom,page11 = <1 0x3b02 2 0x04 1 5>;
+		qcom,pageen11 = <1 0x3b00 2 0x01 1 5>;
+		qcom,poll11 = <1 0x3b01 2 0x01 1 5>;
+		qcom,mem11 = <64 0x3b04 2 0 1 0>;
+
+		qcom,page12 = <1 0x3b02 2 0x05 1 5>;
+		qcom,pageen12 = <1 0x3b00 2 0x01 1 5>;
+		qcom,poll12 = <1 0x3b01 2 0x01 1 5>;
+		qcom,mem12 = <64 0x3b04 2 0 1 0>;
+
+		qcom,page13 = <1 0x3b02 2 0x06 1 5>;
+		qcom,pageen13 = <1 0x3b00 2 0x01 1 5>;
+		qcom,poll13 = <1 0x3b01 2 0x01 1 5>;
+		qcom,mem13 = <64 0x3b04 2 0 1 0>;
+
+		qcom,page14 = <1 0x3b02 2 0x07 1 5>;
+		qcom,pageen14 = <1 0x3b00 2 0x01 1 5>;
+		qcom,poll14 = <1 0x3b01 2 0x01 1 5>;
+		qcom,mem14 = <64 0x3b04 2 0 1 0>;
+
+		qcom,page15 = <1 0x3b02 2 0x08 1 5>;
+		qcom,pageen15 = <1 0x3b00 2 0x01 1 5>;
+		qcom,poll15 = <1 0x3b01 2 0x01 1 5>;
+		qcom,mem15 = <56 0x3b04 2 0 1 0>;
+
+		cam_vdig-supply = <&vph_pwr_vreg>;
+		cam_vio-supply = <&vph_pwr_vreg>;
+		cam_vana-supply = <&vph_pwr_vreg>;
+		qcom,cam-vreg-name = "cam_vdig", "cam_vio", "cam_vana";
+		qcom,cam-vreg-min-voltage = <2100000 0 2850000 2800000>;
+		qcom,cam-vreg-max-voltage = <2100000 0 2850000 2800000>;
+		qcom,cam-vreg-op-mode = <200000 0 80000 100000>;
+		pinctrl-names = "cam_default", "cam_suspend";
+		pinctrl-0 = <&cam_sensor_mclk0_default &cam_sensor_rear_default>;
+		pinctrl-1 = <&cam_sensor_mclk0_sleep &cam_sensor_rear_sleep>;
+		qcom,gpio-no-mux = <0>;
+		gpios = <&msm_gpio 26 0>,
+			<&msm_gpio 35 0>,
+			<&msm_gpio 34 0>;
+		qcom,gpio-reset = <1>;
+		qcom,gpio-standby = <2>;
+		qcom,gpio-req-tbl-num = <0 1 2>;
+		qcom,gpio-req-tbl-flags = <1 0 0>;
+		qcom,gpio-req-tbl-label = "CAMIF_MCLK",
+			"CAM_RESET1",
+			"CAM_STANDBY";
+		qcom,gpio-set-tbl-num = <1 1>;
+		qcom,gpio-set-tbl-flags = <0 2>;
+		qcom,gpio-set-tbl-delay = <1000 4000>;
+		qcom,csi-lane-assign = <0x4320>;
+		qcom,csi-lane-mask = <0x3>;
+		qcom,cam-power-seq-type = "sensor_vreg", "sensor_vreg", "sensor_vreg",
+			"sensor_gpio", "sensor_gpio" , "sensor_clk";
+		qcom,cam-power-seq-val = "cam_vdig", "cam_vana", "cam_vio",
+			"sensor_gpio_reset", "sensor_gpio_standby","sensor_cam_mclk";
+		qcom,cam-power-seq-cfg-val = <1 1 1 1 1 24000000>;
+		qcom,cam-power-seq-delay = <1 1 1 30 30 5>;
+	};
+
+	qcom,camera@78 {
+		compatible = "ovti,ov5645";
+		reg = <0x78 0x0>;
+		qcom,slave-id = <0x78 0x300a 0x5645>;
+		qcom,csiphy-sd-index = <1>;
+		qcom,csid-sd-index = <1>;
+		qcom,mount-angle = <90>;
+		qcom,sensor-name = "ov5645";
+		cam_vdig-supply = <&vph_pwr_vreg>;
+		cam_vana-supply = <&vph_pwr_vreg>;
+		cam_vio-supply = <&vph_pwr_vreg>;
+		qcom,cam-vreg-name = "cam_vdig", "cam_vio", "cam_vana";
+		qcom,cam-vreg-min-voltage = <2100000 0 2850000>;
+		qcom,cam-vreg-max-voltage = <2100000 0 2850000>;
+		qcom,cam-vreg-op-mode = <200000 0 80000>;
+		pinctrl-names = "cam_default", "cam_suspend";
+		pinctrl-0 = <&cam_sensor_mclk1_default &cam_sensor_front_default>;
+		pinctrl-1 = <&cam_sensor_mclk1_sleep &cam_sensor_front_sleep>;
+		gpios = <&msm_gpio 27 0>,
+			<&msm_gpio 28 0>,
+			<&msm_gpio 33 0>;
+		qcom,gpio-reset = <1>;
+		qcom,gpio-standby = <2>;
+		qcom,gpio-req-tbl-num = <0 1 2>;
+		qcom,gpio-req-tbl-flags = <1 0 0>;
+		qcom,gpio-req-tbl-label = "CAMIF_MCLK",
+					"CAM_RESET",
+					"CAM_STANDBY";
+		qcom,gpio-set-tbl-num = <1 1>;
+		qcom,gpio-set-tbl-flags = <0 2>;
+		qcom,gpio-set-tbl-delay = <1000 4000>;
+		qcom,csi-lane-assign = <0x4320>;
+		qcom,csi-lane-mask = <0x3>;
+		qcom,sensor-position = <1>;
+		qcom,sensor-mode = <0>;
+		qcom,cci-master = <0>;
+		qcom,mclk-23880000;
+		clocks = <&clock_gcc clk_mclk1_clk_src>,
+				<&clock_gcc clk_gcc_camss_mclk1_clk>;
+		clock-names = "cam_src_clk", "cam_clk";
+	};
+
+	qcom,camera@0 {
+		cell-index = <0>;
+		compatible = "qcom,camera";
+		reg = <0x0>;
+		qcom,csiphy-sd-index = <0>;
+		qcom,csid-sd-index = <0>;
+		qcom,mount-angle = <90>;
+		qcom,actuator-src = <&actuator0>;
+		qcom,eeprom-src = <&eeprom0>;
+		cam_vdig-supply = <&vph_pwr_vreg>;
+		cam_vana-supply = <&vph_pwr_vreg>;
+		cam_vio-supply = <&vph_pwr_vreg>;
+		cam_vaf-supply = <&vph_pwr_vreg>;
+		qcom,cam-vreg-name = "cam_vdig", "cam_vio", "cam_vana",
+							"cam_vaf";
+		qcom,cam-vreg-min-voltage = <2100000 0 2850000 2800000>;
+		qcom,cam-vreg-max-voltage = <2100000 0 2850000 2800000>;
+		qcom,cam-vreg-op-mode = <200000 0 80000 100000>;
+		pinctrl-names = "cam_default", "cam_suspend";
+		pinctrl-0 = <&cam_sensor_mclk0_default &cam_sensor_rear_default>;
+		pinctrl-1 = <&cam_sensor_mclk0_sleep &cam_sensor_rear_sleep>;
+		gpios = <&msm_gpio 26 0>,
+			<&msm_gpio 35 0>,
+			<&msm_gpio 34 0>;
+		qcom,gpio-reset = <1>;
+		qcom,gpio-standby = <2>;
+		qcom,gpio-req-tbl-num = <0 1 2>;
+		qcom,gpio-req-tbl-flags = <1 0 0>;
+		qcom,gpio-req-tbl-label = "CAMIF_MCLK",
+			"CAM_RESET1",
+			"CAM_STANDBY";
+		qcom,sensor-position = <0>;
+		qcom,sensor-mode = <0>;
+		qcom,cci-master = <0>;
+		qcom,mclk-23880000;
+		status = "ok";
+		clocks = <&clock_gcc clk_mclk0_clk_src>,
+				<&clock_gcc clk_gcc_camss_mclk0_clk>;
+		clock-names = "cam_src_clk", "cam_clk";
+	};
+
+	qcom,camera@1 {
+		cell-index = <1>;
+		compatible = "qcom,camera";
+		reg = <0x1>;
+		qcom,csiphy-sd-index = <1>;
+		qcom,csid-sd-index = <1>;
+		qcom,mount-angle = <90>;
+		cam_vdig-supply = <&vph_pwr_vreg>;
+		cam_vana-supply = <&vph_pwr_vreg>;
+		cam_vio-supply = <&vph_pwr_vreg>;
+		qcom,cam-vreg-name = "cam_vdig", "cam_vio", "cam_vana";
+		qcom,cam-vreg-min-voltage = <2100000 0 2850000>;
+		qcom,cam-vreg-max-voltage = <2100000 0 2850000>;
+		qcom,cam-vreg-op-mode = <200000 0 80000>;
+		pinctrl-names = "cam_default", "cam_suspend";
+		pinctrl-0 = <&cam_sensor_mclk1_default &cam_sensor_front_default>;
+		pinctrl-1 = <&cam_sensor_mclk1_sleep &cam_sensor_front_sleep>;
+		gpios = <&msm_gpio 27 0>,
+			<&msm_gpio 28 0>,
+			<&msm_gpio 33 0>;
+		qcom,gpio-reset = <1>;
+		qcom,gpio-standby = <2>;
+		qcom,gpio-req-tbl-num = <0 1 2>;
+		qcom,gpio-req-tbl-flags = <1 0 0>;
+		qcom,gpio-req-tbl-label = "CAMIF_MCLK",
+			"CAM_RESET",
+			"CAM_STANDBY";
+		qcom,cci-master = <0>;
+		qcom,mclk-23880000;
+		status = "ok";
+		clocks = <&clock_gcc clk_mclk1_clk_src>,
+				<&clock_gcc clk_gcc_camss_mclk1_clk>;
+		clock-names = "cam_src_clk", "cam_clk";
+	};
+};
diff --git a/arch/arm/boot/dts/qcom/apq8016-sbc.dts b/arch/arm/boot/dts/qcom/apq8016-sbc.dts
new file mode 100644
index 0000000..27387fe
--- /dev/null
+++ b/arch/arm/boot/dts/qcom/apq8016-sbc.dts
@@ -0,0 +1,25 @@
+/* Copyright (c) 2015, The Linux Foundation. All rights reserved.
+*
+* This program is free software; you can redistribute it and/or modify
+* it under the terms of the GNU General Public License version 2 and
+* only version 2 as published by the Free Software Foundation.
+*
+* This program is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+* GNU General Public License for more details.
+*/
+
+/dts-v1/;
+
+#include "apq8016.dtsi"
+#include "apq8016-sbc.dtsi"
+
+
+/ {
+	model = "Qualcomm Technologies, Inc. APQ 8016 SBC";
+	compatible = "qcom,apq8016", "qcom,sbc";
+	qcom,msm-id = <247 0>;
+	qcom,board-id = <24 0>;
+
+};
diff --git a/arch/arm/boot/dts/qcom/apq8016-sbc.dtsi b/arch/arm/boot/dts/qcom/apq8016-sbc.dtsi
new file mode 100644
index 0000000..07599c0
--- /dev/null
+++ b/arch/arm/boot/dts/qcom/apq8016-sbc.dtsi
@@ -0,0 +1,663 @@
+/* Copyright (c) 2015, The Linux Foundation. All rights reserved.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 and
+ * only version 2 as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ */
+
+#include "msm8916.dtsi"
+#include "msm8916-pinctrl.dtsi"
+#include "apq8016-camera-sensor-sbc.dtsi"
+#include "dsi-panel-jdi-1080p-video.dtsi"
+
+/ {
+	aliases {
+		serial0 = &blsp1_uart2;
+	};
+};
+
+&soc {
+	i2c@78b6000 { /* BLSP1 QUP2 */
+		mpu6050@68 { /* Gyroscope and accelerometer sensor combo */
+			compatible = "invn,mpu6050";
+			reg = <0x68>;
+			pinctrl-names = "mpu_default","mpu_sleep";
+			pinctrl-0 = <&mpu6050_default>;
+			pinctrl-1 = <&mpu6050_sleep>;
+			interrupt-parent = <&msm_gpio>;
+			interrupts = <115 0x2>;
+			vdd-supply = <&vph_pwr_vreg>;
+			vlogic-supply = <&vph_pwr_vreg>;
+			vi2c-supply = <&vph_pwr_vreg>;
+			invn,gpio-int = <&msm_gpio 115 0x2>;
+			invn,place = "Portrait Down Back Side";
+		};
+
+		avago@39 { /* Ambient light and proximity sensor */
+			compatible = "avago,apds9900";
+			reg = <0x39>;
+			pinctrl-names = "default","sleep";
+			pinctrl-0 = <&apds99xx_default>;
+			pinctrl-1 = <&apds99xx_sleep>;
+			interrupt-parent = <&msm_gpio>;
+			interrupts = <113 0x2002>;
+			vdd-supply = <&vph_pwr_vreg>;
+			vio-supply = <&vph_pwr_vreg>;
+			avago,irq-gpio = <&msm_gpio 113 0x2002>;
+			avago,ps-threshold = <600>;
+			avago,ps-hysteresis-threshold = <500>;
+			avago,ps-pulse = <8>;
+			avago,ps-pgain = <0>;
+			avago,als-B = <223>;
+			avago,als-C = <70>;
+			avago,als-D = <142>;
+			avago,ga-value = <48>;
+		};
+
+		akm@c {
+			compatible = "ak,ak8963";
+			reg = <0x0c>;
+			pinctrl-names = "ak8963_default", "ak8963_sleep";
+			pinctrl-0 = <&ak8963_default>;
+			pinctrl-1 = <&ak8963_sleep>;
+			interrupt-parent = <&msm_gpio>;
+			interrupts = <69 0x2>;
+			vdd-supply = <&vph_pwr_vreg>;
+			vio-supply = <&vph_pwr_vreg>;
+			ak,layout = <0x6>;
+			ak,auto-report;
+		};
+	};
+
+    spi_0 { /* BLSP1 QUP3 */
+		status = "disabled";
+    };
+
+	i2c@78b8000 { /* BLSP1 QUP4 */
+       /*DSI_TO_HDMI insert I2C configuration */
+	};
+
+	i2c@78b9000 { /* BLSP1 QUP5 */
+		synaptics@20 {
+			compatible = "synaptics,rmi4";
+			reg = <0x20>;
+			interrupt-parent = <&msm_gpio>;
+			interrupts = <13 0x2008>;
+			vdd-supply = <&vph_pwr_vreg>;
+			vcc_i2c-supply = <&vph_pwr_vreg>;
+			/* pins used by touchscreen */
+			pinctrl-names = "pmx_ts_active","pmx_ts_suspend","pmx_ts_suspend";
+			pinctrl-0 = <&ts_int_active &ts_reset_active>;
+			pinctrl-1 = <&ts_int_suspend &ts_reset_suspend>;
+			pinctrl-2 = <&ts_release>;
+			synaptics,display-coords = <0 0 1079 1919>;
+			synaptics,panel-coords = <0 0 1079 2084>;
+			synaptics,irq-gpio = <&msm_gpio 13 0x2008>;
+			synaptics,reset-gpio = <&msm_gpio 12 0x0>;
+			synaptics,i2c-pull-up;
+			synaptics,power-down;
+			synaptics,disable-gpios;
+			/* Underlying clocks used by secure touch */
+			clocks = <&clock_gcc clk_gcc_blsp1_ahb_clk>,
+			 <&clock_gcc clk_gcc_blsp1_qup5_i2c_apps_clk>;
+			clock-names = "iface_clk", "core_clk";
+		};
+	};
+
+	i2c@78ba000 { /* BLSP1 QUP6 */
+        /*insert i2c_6 states expansion connector  */
+	};
+
+	/*
+	* vph_pwr_vreg represents the unregulated battery voltage supply
+	* VPH_PWR that is present whenever the device is powered on.
+	*/
+	vph_pwr_vreg: vph_pwr_vreg {
+		compatible = "regulator-fixed";
+		regulator-name = "vph_pwr";
+		status = "ok";
+		regulator-always-on;
+		enable-active-high;
+	};
+
+
+	gen-vkeys {
+		compatible = "qcom,gen-vkeys";
+		label = "synaptics_rmi4_i2c";
+		qcom,disp-maxx = <1079>;
+		qcom,disp-maxy = <1919>;
+		qcom,panel-maxx = <1079>;
+		qcom,panel-maxy = <2084>;
+		qcom,key-codes = <158 139 172 217>;
+	};
+
+
+	sound {
+		compatible = "qcom,msm8x16-audio-codec";
+		qcom,model = "msm8x16-snd-card-sbc";
+		qcom,msm-snd-card-id = <0>;
+		qcom,msm-codec-type = "internal";
+		qcom,msm-ext-pa = "primary";
+		qcom,msm-mclk-freq = <9600000>;
+		qcom,msm-mbhc-hphl-swh = <0>;
+		qcom,msm-mbhc-gnd-swh = <0>;
+		qcom,msm-micbias1-ext-cap;
+		qcom,msm-hs-micbias-type = "internal";
+		qcom,audio-routing =
+			"RX_BIAS", "MCLK",
+			"SPK_RX_BIAS", "MCLK",
+			"INT_LDO_H", "MCLK",
+			"MIC BIAS External", "Handset Mic",
+			"MIC BIAS Internal2", "Headset Mic",
+			"MIC BIAS External", "Secondary Mic",
+			"AMIC1", "MIC BIAS External",
+			"AMIC2", "MIC BIAS Internal2",
+			"AMIC3", "MIC BIAS External",
+			"DMIC1", "MIC BIAS Internal1",
+			"MIC BIAS Internal1", "Digital Mic1",
+			"DMIC2", "MIC BIAS Internal1",
+			"MIC BIAS Internal1", "Digital Mic2";
+		pinctrl-names = "cdc_lines_act",
+				"cdc_lines_sus",
+				"cdc_lines_sec_ext_act",
+				"cdc_lines_sec_ext_sus",
+				"cross_conn_det_act",
+				"cross_conn_det_sus";
+		pinctrl-0 = <&cdc_pdm_lines_act>;
+		pinctrl-1 = <&cdc_pdm_lines_sus>;
+		pinctrl-2 = <&cdc_pdm_lines_act &cdc_ext_pa_act
+						&cdc_ext_pa_ws_act>;
+		pinctrl-3 = <&cdc_pdm_lines_sus &cdc_ext_pa_sus
+						&cdc_ext_pa_ws_sus>;
+		pinctrl-4 = <&cross_conn_det_act>;
+		pinctrl-5 = <&cross_conn_det_sus>;
+
+		asoc-platform = <&pcm0>, <&pcm1>, <&voip>, <&voice>,
+				<&loopback>, <&compress>, <&hostless>,
+				<&afe>, <&lsm>, <&routing>, <&lpa>;
+		asoc-platform-names = "msm-pcm-dsp.0", "msm-pcm-dsp.1",
+				"msm-voip-dsp", "msm-pcm-voice", "msm-pcm-loopback",
+				"msm-compress-dsp", "msm-pcm-hostless", "msm-pcm-afe",
+				"msm-lsm-client", "msm-pcm-routing", "msm-pcm-lpa";
+		asoc-cpu = <&dai_pri_auxpcm>, <&dai_hdmi>,
+				<&dai_mi2s0>, <&dai_mi2s1>, <&dai_mi2s2>, <&dai_mi2s3>,
+				<&sb_0_rx>, <&sb_0_tx>, <&sb_1_rx>, <&sb_1_tx>,
+				<&sb_3_rx>, <&sb_3_tx>, <&sb_4_rx>, <&sb_4_tx>,
+				<&bt_sco_rx>, <&bt_sco_tx>, <&int_fm_rx>, <&int_fm_tx>,
+				<&afe_pcm_rx>, <&afe_pcm_tx>, <&afe_proxy_rx>, <&afe_proxy_tx>,
+				<&incall_record_rx>, <&incall_record_tx>, <&incall_music_rx>,
+				<&incall_music_2_rx>;
+		asoc-cpu-names = "msm-dai-q6-auxpcm.1", "msm-dai-q6-hdmi.8",
+				"msm-dai-q6-mi2s.0", "msm-dai-q6-mi2s.1",
+				"msm-dai-q6-mi2s.2", "msm-dai-q6-mi2s.3",
+				"msm-dai-q6-dev.16384", "msm-dai-q6-dev.16385",
+				"msm-dai-q6-dev.16386", "msm-dai-q6-dev.16387",
+				"msm-dai-q6-dev.16390", "msm-dai-q6-dev.16391",
+				"msm-dai-q6-dev.16392", "msm-dai-q6-dev.16393",
+				"msm-dai-q6-dev.12288", "msm-dai-q6-dev.12289",
+				"msm-dai-q6-dev.12292", "msm-dai-q6-dev.12293",
+				"msm-dai-q6-dev.224", "msm-dai-q6-dev.225",
+				"msm-dai-q6-dev.241", "msm-dai-q6-dev.240",
+				"msm-dai-q6-dev.32771", "msm-dai-q6-dev.32772",
+				"msm-dai-q6-dev.32773", "msm-dai-q6-dev.32770";
+		asoc-codec = <&stub_codec>, <&pm8916_tombak_dig>;
+		asoc-codec-names = "msm-stub-codec.1", "tombak_codec";
+	};
+
+	gpio-leds {
+		compatible = "gpio-leds";
+		status = "okay";
+		pinctrl-names = "default";
+		pinctrl-0 = <&gpio_led_off>;
+
+		general1 {
+			gpios = <&msm_gpio 21 0>;
+			label = "led1";
+			linux,default-trigger = "none";
+			default-state = "off";
+			retain-state-suspended;
+		};
+
+		general2 {
+			gpios = <&msm_gpio 120 0>;
+			label = "led2";
+			linux,default-trigger = "none";
+			default-state = "off";
+			retain-state-suspended;
+		};
+	};
+
+	gpio_keys {
+		compatible = "gpio-keys";
+		input-name = "gpio-keys";
+		pinctrl-names = "tlmm_gpio_key_active","tlmm_gpio_key_suspend";
+		pinctrl-0 = <&gpio_key_active>;
+		pinctrl-1 = <&gpio_key_suspend>;
+
+		camera_focus {
+			label = "camera_focus";
+			gpios = <&msm_gpio 108 0x1>;
+			linux,input-type = <1>;
+			linux,code = <0x210>;
+			gpio-key,wakeup;
+			debounce-interval = <15>;
+		};
+
+		camera_snapshot {
+			label = "camera_snapshot";
+			gpios = <&msm_gpio 109 0x1>;
+			linux,input-type = <1>;
+			linux,code = <0x2fe>;
+			gpio-key,wakeup;
+			debounce-interval = <15>;
+		};
+
+		vol_up {
+			label = "volume_up";
+			gpios = <&msm_gpio 107 0x1>;
+			linux,input-type = <1>;
+			linux,code = <115>;
+			gpio-key,wakeup;
+			debounce-interval = <15>;
+		};
+	};
+};
+
+&usb_otg {
+	qcom,hsusb-otg-mode = <3>;
+	qcom,usbid-gpio = <&msm_gpio 121 0>;
+	pinctrl-names = "default";
+	pinctrl-0 = <&usbid_default>;
+	vbus_otg-supply = <&vph_pwr_vreg>;
+};
+
+&blsp1_uart1 {
+	status = "ok";
+};
+
+&blsp1_uart2 {
+	status = "ok";
+	pinctrl-names = "default";
+	pinctrl-0 = <&uart_console_sleep>;
+};
+
+&sdhc_1 {
+	vdd-supply = <&pm8916_l8>;
+	qcom,vdd-voltage-level = <2900000 2900000>;
+	qcom,vdd-current-level = <200 400000>;
+
+	vdd-io-supply = <&pm8916_l5>;
+	qcom,vdd-io-always-on;
+	qcom,vdd-io-lpm-sup;
+	qcom,vdd-io-voltage-level = <1800000 1800000>;
+	qcom,vdd-io-current-level = <200 60000>;
+
+	pinctrl-names = "active", "sleep";
+	pinctrl-0 = <&sdc1_clk_on &sdc1_cmd_on &sdc1_data_on>;
+	pinctrl-1 = <&sdc1_clk_off &sdc1_cmd_off &sdc1_data_off>;
+
+	qcom,nonremovable;
+
+	status = "ok";
+};
+
+&sdhc_2 {
+	vdd-supply = <&pm8916_l11>;
+	qcom,vdd-voltage-level = <2800000 2950000>;
+	qcom,vdd-current-level = <15000 400000>;
+
+	vdd-io-supply = <&pm8916_l12>;
+	qcom,vdd-io-voltage-level = <1800000 2950000>;
+	qcom,vdd-io-current-level = <200 50000>;
+
+	pinctrl-names = "active", "sleep";
+	pinctrl-0 = <&sdc2_clk_on &sdc2_cmd_on &sdc2_data_on &sdc2_cd_on>;
+	pinctrl-1 = <&sdc2_clk_off &sdc2_cmd_off &sdc2_data_off &sdc2_cd_off>;
+
+	#address-cells = <0>;
+	interrupt-parent = <&sdhc_2>;
+	interrupts = <0 1 2>;
+	#interrupt-cells = <1>;
+	interrupt-map-mask = <0xffffffff>;
+	interrupt-map = <0 &intc 0 125 0
+			1 &intc 0 221 0
+			2 &msm_gpio 38 0>;
+	interrupt-names = "hc_irq", "pwr_irq", "status_irq";
+	cd-gpios = <&msm_gpio 38 0x1>;
+
+	status = "ok";
+};
+
+&mdss_mdp {
+	qcom,mdss-pref-prim-intf = "dsi";
+};
+
+&dsi_jdi_1080_vid {
+	qcom,mdss-dsi-bl-pmic-control-type = "bl_ctrl_pwm";
+	qcom,mdss-dsi-bl-pmic-pwm-frequency = <100>;
+	qcom,mdss-dsi-bl-pmic-bank-select = <0>;
+	qcom,mdss-dsi-pwm-gpio = <&pm8916_mpps 4 0>;
+
+};
+
+&mdss_dsi0 {
+	qcom,dsi-pref-prim-pan = <&dsi_jdi_1080_vid>;
+	pinctrl-names = "mdss_default", "mdss_sleep";
+	pinctrl-0 = <&mdss_dsi_active &mdss_te_active>;
+	pinctrl-1 = <&mdss_dsi_suspend &mdss_te_suspend>;
+
+	qcom,platform-reset-gpio = <&msm_gpio 25 0>;
+};
+
+&dsi_jdi_1080_vid {
+       qcom,cont-splash-enabled;
+};
+
+&qcom_tzlog {
+	status = "okay";
+};
+
+&qcom_rng {
+	status = "okay";
+};
+
+&qcom_crypto {
+	status = "okay";
+};
+
+&qcom_cedev {
+	status = "okay";
+};
+
+&qcom_seecom {
+	status = "okay";
+};
+
+/* CoreSight */
+&tpiu {
+      pinctrl-names = "sdcard", "trace", "swduart",
+		      "swdtrc", "jtag", "spmi";
+       /* NIDnT */
+      pinctrl-0 = <&qdsd_clk_sdcard &qdsd_cmd_sdcard
+                  &qdsd_data0_sdcard &qdsd_data1_sdcard
+                  &qdsd_data2_sdcard &qdsd_data3_sdcard>;
+      pinctrl-1 = <&qdsd_clk_trace &qdsd_cmd_trace
+                  &qdsd_data0_trace &qdsd_data1_trace
+                  &qdsd_data2_trace &qdsd_data3_trace>;
+      pinctrl-2 = <&qdsd_cmd_swduart &qdsd_data0_swduart
+                  &qdsd_data1_swduart &qdsd_data2_swduart
+                  &qdsd_data3_swduart>;
+      pinctrl-3 = <&qdsd_clk_swdtrc &qdsd_cmd_swdtrc
+                  &qdsd_data0_swdtrc &qdsd_data1_swdtrc
+                  &qdsd_data2_swdtrc &qdsd_data3_swdtrc>;
+      pinctrl-4 = <&qdsd_cmd_jtag &qdsd_data0_jtag
+                  &qdsd_data1_jtag &qdsd_data2_jtag
+                  &qdsd_data3_jtag>;
+      pinctrl-5 = <&qdsd_clk_spmi &qdsd_cmd_spmi
+                  &qdsd_data0_spmi &qdsd_data3_spmi>;
+};
+
+&pm8916_chg {
+        status = "ok";
+        qcom,charging-disabled;
+        qcom,disable-follow-on-reset;
+};
+
+&pm8916_bms {
+        status = "ok";
+        qcom,disable-bms;
+};
+
+&tlmm_pinmux {
+	usb-id-pin {
+		qcom,pins = <&gp 121>;
+	     usbid_default{
+			drive-strength = <2>;
+			bias-disable; /* no pullup */
+			output-low;
+		};
+	};
+
+	gpio_led_pins {
+		qcom,pins = <&gp 21>, <&gp 120>;
+		qcom,num-grp-pins = <2>;
+		label = "gpio-led-pins";
+		gpio_led_off: led_off {
+			drive-strength = <2>;
+			bias-disable; /* no pullup */
+			output-low;
+		};
+	};
+
+	ice40-spi-usb-pins {
+		status = "disabled";
+	};
+
+	cam_sensor_flash {
+		status = "disabled";
+    };
+
+	pmx_rd_nfc_int {
+		status = "disabled";
+    };
+
+	pmx_nfc_reset {
+		status = "disabled";
+    };
+
+    pmx_i2c_5 {
+		status = "disabled";
+    };
+
+    pmx_mdss_te {
+	    qcom,num-grp-pins = <1>;
+	    qcom,pins = <&gp 24>;
+    };
+
+    pmx_mdss {
+	    qcom,num-grp-pins = <1>;
+	    qcom,pins = <&gp 25>;
+    };
+
+    tpiu_setb_9 {
+       status = "disabled";
+    };
+
+};
+
+
+&pm8916_mpps {
+	mpp@a000 { /* MPP 1 */
+		/* VDD_PX */
+		status = "disabled";
+	};
+
+	mpp@a100 { /* MPP 2 */
+		/* HR LED */
+	};
+
+	mpp@a200 { /* MPP 3 */
+		/* VREF DAC */
+		status = "disabled";
+	};
+
+	mpp@a300 { /* MPP 4 */
+		/* Backlight PWM */
+		qcom,mode = <1>;		/* Digital output */
+		qcom,invert = <0>;		/* Disable invert */
+		qcom,src-sel = <4>;		/* DTEST1 */
+		qcom,vin-sel = <0>;		/* VPH_PWR */
+		qcom,master-en = <1>;		/* Enable MPP */
+	};
+};
+
+&pm8916_gpios {
+	/* USB_HUB_RESET_N*/
+	gpio@c000 { /* GPIO 1 */
+		qcom,mode = <0>;	/* DIGITAL IN */
+		qcom,pull = <2>;	/* PULL UP */
+		qcom,vin-sel = <3>;	/* VPH */
+		qcom,src-sel = <0>;	/* CONSTANT */
+		qcom,master-en = <1>; 	/* ENABLE GPIO */
+		status = "okay";
+	};
+
+	/* USB_SW_SEL*/
+	gpio@c100 { /* GPIO 2 */
+		qcom,mode = <0>;	/* DIGITAL IN */
+		qcom,pull = <5>;	/* No PULL */
+		qcom,vin-sel = <0>;	/* VPH */
+		qcom,src-sel = <0>;	/* CONSTANT */
+		qcom,master-en = <1>; 	/* ENABLE GPIO */
+		status = "okay";
+	};
+
+	gpio@c200 { /* GPIO 3 */
+		qcom,mode = <1>;		/* Digital output */
+		qcom,pull = <4>;		/* Pull down */
+		qcom,vin-sel = <2>;		/* VIN 2*/
+		qcom,src-sel = <1>;		/* GPIO */
+		qcom,master-en = <1>;	/* Enable GPIO */
+		status = "okay";
+	};
+
+	gpio@c300 { /* GPIO 4 */
+		qcom,mode = <1>;		/* Digital output */
+		qcom,pull = <4>;		/* Pull down */
+		qcom,vin-sel = <2>;		/* VIN 2*/
+		qcom,src-sel = <1>;		/* GPIO */
+		qcom,master-en = <1>;	/* Enable GPIO */
+		status = "okay";
+	};
+};
+
+&spmi_bus {
+	qcom,pm8916@0 {
+		qcom,leds@c200  {
+			compatible = "qcom,leds-qpnp";
+			reg = <0xc200 0x100>;
+			status = "okay";
+			qcom,led_gpio_3 {
+				label = "gpio";
+				linux,name = "led3";
+				qcom,max-current = <40>;
+				qcom,id = <8>;
+				linux,default-trigger = "none";
+				qcom,default-state = "off";
+				qcom,source-sel = <1>;
+				qcom,mode-ctrl = <0x10>;
+				qcom,vin-ctrl = <0x02>;
+			};
+		};
+		qcom,leds@c300  {
+			compatible = "qcom,leds-qpnp";
+			reg = <0xc300 0x100>;
+			status = "okay";
+			qcom,led_gpio_4 {
+				label = "gpio";
+				linux,name = "boot";
+				qcom,max-current = <40>;
+				qcom,id = <8>;
+				linux,default-trigger =
+						"none";
+				qcom,default-state = "on";
+				qcom,source-sel = <1>;
+				qcom,mode-ctrl = <0x10>;
+				qcom,vin-ctrl = <0x02>;
+			};
+		};
+	};
+};
+
+/* RPM controlled regulators */
+&rpm_bus {
+	rpm-regulator-ldoa4 {
+		status = "disabled";
+		pm8916_l4 {
+			status = "disabled";
+		};
+	};
+
+	rpm-regulator-ldoa10 {
+		status = "disabled";
+		pm8916_l10 {
+			status = "disabled";
+		};
+	};
+
+	rpm-regulator-ldoa14 {
+		status = "disabled";
+		pm8916_l14 {
+			status = "disabled";
+		};
+	};
+
+	rpm-regulator-ldoa15 {
+		pm8916_l15 {
+			regulator-always-on;
+		};
+	};
+
+	rpm-regulator-ldoa16 {
+		pm8916_l16{
+			regulator-always-on;
+		};
+	};
+
+	rpm-regulator-ldoa17 {
+		pm8916_l17 {
+			regulator-min-microvolt = <3300000>;
+			regulator-max-microvolt = <3300000>;
+			qcom,init-voltage = <3300000>;
+		};
+	};
+
+	rpm-regulator-ldoa18 {
+		status = "disabled";
+		pm8916_l18 {
+			status = "disabled";
+		};
+	};
+
+	rpm-regulator-smpa3 {
+		pm8916_s3{
+			qcom,init-voltage = <1300000>;
+		};
+	};
+
+};
+
+&spmi_bus {
+	qcom,pm8916@0 {
+		qcom,leds@a100 { /* WLAN LED */
+		status = "okay";
+			qcom,led_mpp_2 {
+				label = "mpp";
+				linux,name = "wlan";
+				qcom,max-current = <40>;
+				qcom,id = <6>;
+				qcom,source-sel = <1>;
+				qcom,mode-ctrl = <0x60>;
+				qcom,mode = "manual";
+				linux,default-trigger =
+					"none";
+			};
+		};
+	};
+
+	qcom,pm8916@1 {
+		qcom,vibrator@c000 {
+			status = "okay";
+			qcom,vib-timeout-ms = <15000>;
+			qcom,vib-vtg-level-mV = <3100>;
+		};
+	};
+};
+
+
+
diff --git a/arch/arm/boot/dts/qcom/apq8016.dtsi b/arch/arm/boot/dts/qcom/apq8016.dtsi
new file mode 100644
index 0000000..7fa77b4
--- /dev/null
+++ b/arch/arm/boot/dts/qcom/apq8016.dtsi
@@ -0,0 +1,33 @@
+/* Copyright (c) 2015, The Linux Foundation. All rights reserved.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 and
+ * only version 2 as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ */
+
+/*
+ * As a general rule, only version-specific property overrides should be placed
+ * inside this file. However, device definitions should be placed inside the
+ * 8916.dtsi file.
+ */
+
+#include "msm8916.dtsi"
+#include "msm8916-memory.dtsi"
+#include "msm8916-pinctrl.dtsi"
+
+/{
+	memory {
+		modem_adsp_mem: modem_adsp_region@0 {
+			reg = <0x0 0x86800000 0x0 0x02b00000>;
+		};
+
+		peripheral_mem: pheripheral_region@0 {
+			reg = <0x0 0x89300000 0x0 0x0600000>;
+		};
+	};
+};
diff --git a/arch/arm/boot/dts/qcom/msm8916-pinctrl.dtsi b/arch/arm/boot/dts/qcom/msm8916-pinctrl.dtsi
index 412d519..304a4ed 100644
--- a/arch/arm/boot/dts/qcom/msm8916-pinctrl.dtsi
+++ b/arch/arm/boot/dts/qcom/msm8916-pinctrl.dtsi
@@ -1,4 +1,4 @@
-/* Copyright (c) 2013-2014, The Linux Foundation. All rights reserved.
+/* Copyright (c) 2013-2015, The Linux Foundation. All rights reserved.
  *
  * This program is free software; you can redistribute it and/or modify
  * it under the terms of the GNU General Public License version 2 and
@@ -344,6 +344,58 @@
 			};
 		};
 
+		spi2_active {
+			/* MOSI, MISO, CLK */
+			qcom,pins = <&gp 16>, <&gp 17>, <&gp 19>;
+			qcom,num-grp-pins = <3>;
+			qcom,pin-func = <1>;
+			label = "spi2-active";
+			/* active state */
+			spi2_default: default {
+				drive-strength = <12>; /* 12 MA */
+				bias-disable = <0>; /* No PULL */
+			};
+		};
+
+		spi2_suspend {
+			/* MOSI, MISO, CLK */
+			qcom,pins = <&gp 16>, <&gp 17>, <&gp 19>;
+			qcom,num-grp-pins = <3>;
+			qcom,pin-func = <0>;
+			label = "spi2-suspend";
+			/* suspended state */
+			spi2_sleep: sleep {
+				drive-strength = <2>; /* 2 MA */
+				bias-pull-down; /* pull down */
+			};
+		};
+		spi2_cs2_active {
+			/* CS */
+			qcom,pins = <&gp 18>;
+			qcom,num-grp-pins = <1>;
+			qcom,pin-func = <1>;
+			label = "spi2-cs0-active";
+			spi2_cs2_active: cs2_active {
+				drive-strength = <2>;
+				bias-disable = <0>;
+			};
+		};
+
+
+		spi2_cs2_suspend {
+			/* CS */
+			qcom,pins = <&gp 18>;
+			qcom,num-grp-pins = <1>;
+			qcom,pin-func = <0>;
+			label = "spi2-cs0-suspend";
+			spi2_cs2_sleep: cs2_sleep {
+				drive-strength = <2>;
+				bias-disable = <0>;
+			};
+		};
+
+
+
 		ice40-spi-usb-pins {
 			qcom,pins = <&gp 0>, <&gp 1>, <&gp 3>, <&gp 114>;
 			qcom,num-grp-pins = <4>;
-- 
1.8.2.1

